ToolBox/CoinApi/CoinApiSymbolMapper.cs

```diff
@@ -0,0 +1,178 @@
+        /// </summary>
+        /// <param name="symbol">A Lean symbol instance</param>
+        /// <returns>The CoinAPI symbol id</returns>
+        public string GetBrokerageSymbol(Symbol symbol)
+        {
+            string symbolId;
+            if (!_symbolMap.TryGetValue(symbol, out symbolId))
+            {
+                throw new Exception($"CoinApiSymbolMapper.GetBrokerageSymbol(): Symbol not found: {symbol}");
+            }
+
+            return symbolId;
+        }
+
+        /// <summary>
+        /// Converts a CoinAPI symbol id to a Lean symbol instance
+        /// </summary>
+        /// <param name="brokerageSymbol">The CoinAPI symbol id</param>
+        /// <param name="securityType">The security type</param>
+        /// <param name="market">The market</param>
+        /// <param name="expirationDate">Expiration date of the security (if applicable)</param>
+        /// <param name="strike">The strike of the security (if applicable)</param>
+        /// <param name="optionRight">The option right of the security (if applicable)</param>
+        /// <returns>A new Lean Symbol instance</returns>
+        public Symbol GetLeanSymbol(string brokerageSymbol, SecurityType securityType, string market,
+            DateTime expirationDate = new DateTime(), decimal strike = 0, OptionRight optionRight = OptionRight.Call)
+        {
+            var parts = brokerageSymbol.Split('_');
+            if (parts.Length != 4 || parts[1] != "SPOT")
+            {
+                throw new Exception($"CoinApiSymbolMapper.GetLeanSymbol(): Unsupported SymbolId: {brokerageSymbol}");
+            }
+
+            string symbolMarket;
+            if (!MapExchangeIdsToMarkets.TryGetValue(parts[0], out symbolMarket))
+            {
+                throw new Exception($"CoinApiSymbolMapper.GetLeanSymbol(): Unsupported ExchangeId: {parts[0]}");
+            }
+
+            var baseCurrency = ConvertCoinApiCurrencyToLeanCurrency(parts[2], symbolMarket);
+            var quoteCurrency = ConvertCoinApiCurrencyToLeanCurrency(parts[3], symbolMarket);
+
+            var ticker = baseCurrency + quoteCurrency;
+
+            return Symbol.Create(ticker, SecurityType.Crypto, symbolMarket);
+        }
+
+        /// <summary>
+        /// Returns the CoinAPI exchange id for the given market
+        /// </summary>
+        /// <param name="market">The Lean market</param>
+        /// <returns>The CoinAPI exchange id</returns>
+        public string GetExchangeId(string market)
+        {
+            string exchangeId;
+            MapMarketsToExchangeIds.TryGetValue(market, out exchangeId);
+
+            return exchangeId;
+        }
+
+        private void LoadSymbolMap(string[] exchangeIds)
+        {
+            var list = string.Join(",", exchangeIds);
+
+            using (var wc = new WebClient())
+            {
+                var url = $"{RestUrl}/v1/symbols?filter_symbol_id={list}&apiKey={_apiKey}";
```

Comment:
For future design consideration, this would have been better without the external dependencies. This causes rate limits when applied massively in parallel. 

The pattern for this problem in LEAN is the `HashSet<string> KnownSymbolStrings = new HashSet<string>`. The pattern isn't perfect, but it also doesn't fail if you run it 100x =)